### ISSUE TITLE
CSP: Specify response checking.

### ISSUE THREAD

**Issue by mikewest** [w3c/webappsec#227]
From a thread with @annevk, @sicking, and friends:

> The question here is what "if it matches" means. Keep in mind that
when the SW loads the request, it often won't know that it's doing a
load for an image. The SW might just do a plain fetch() which to the
platform looks unconnected to the <img> request. In fact, that's going
to be a common case given that SW will load a lot of resources during
installation.

Right. In that case, the fetch would be governed by the Service Worker's own CSP. The response would be gated by the page's CSP based on the response type. That is, given a document on example.com with `default-src 'self';` and a Service Worker with `default-src 'self' not-example.com`:

1. The page requests `example.com/image.png`.
2. In Fetch, this matches the document's CSP, so heads on over to the Service Worker.
3. The Service Worker fetches `not-example.com/image.png`, and uses that as the response.
4. At the bottom of Fetch, this response doesn't match the document's CSP, so it's blocked.

We should really specify step 4.

**Comment by annevk** [comment]
@sicking proposed a model whereby a SW cannot regress the CSP of a Client. For that model to work a `Response` instance needs to have an internal slot containing its redirect chain.

That way when a Client fetches a resource and SW invokes `fetch()`, which has several redirects, and returns the response, Client can still use its CSP policy against all URLs that were used to obtain the response.

My proposal would be to add a field "originating urls" to responses to track this. It won't be exposed to script, but CSP can use it. Something like this field is needed also for the simple case where response's url is overwritten by request's url.

**Comment by wanderview** [comment]
Another question I ran into recently was "should synthetic Responses" match CSP "self"?

I believe @sicking was suggesting it should not so that the CSP rule is not diluted.

**Comment by sicking** [comment]
A synthetic Response for a <script src="..."> seems pretty equivalent to a call to eval(), so anytime eval() is forbidden, it seems like we should forbid using a synthetic Response for script loads.

**Comment by annevk** [comment]
Yup. Pretty sure @mikewest agrees. Just needs to be written down...

**Comment by mikewest** [comment]
Mike does agree, he's behind on everything. But it's on my list for this quarter, so, yay?

**Comment by wanderview** [comment]
Does chrome do anything like this yet in its currently shipping code?

**Comment by mikewest** [comment]
No.
